src/toxcore/dht/server.rs

```diff
@@ -124,14 +129,24 @@ impl Server {
             None
         }
     }
+    /// remove timed-out clients
+    pub fn remove_timedout_clients(&self, timeout: Duration) -> IoFuture<()> {
+        let mut state = self.state.write();
+        state.peers_cache.retain(|&_pk, ref client|
+            client.last_resp_time.elapsed() <= timeout);
+        Box::new(future::ok(()))
+    }
     /// send PingRequest to all peer in kbucket    
     pub fn send_pings(&self) -> IoFuture<()> {
-        let mut state = self.state.write();
```

Comment:
Last PR you wrote this code: https://github.com/tox-rs/tox/commit/992eceb89db3b7b995cb13e1210a962c83825ae6#diff-38c05bf285637be9e25083096c19d4e0L128

Please revert it, don't make these changes in the current PR. 1 write lock per method will be fine

Reply:
For remove timed-out clients, create_client() has been changed.
The change is from (just create client) to (check if it exists in peers_cache).
If client already exists, it means that PingRequest() is sent and client had been created and inserted to peers_cache.
Because ping period is 60 seconds and time-out period is 122 seconds.
If every PingRequest create new client, then new client replace old client with same PK.
To do this, change happen in create_client() and to avoid deadlock send_pings() logic is changed.

Reply:
there will be no deadlocks since `write` method waits untill the resourses are released and are free to be used

Reply:
Please try to implement it with 1 write lock since calling to `.write()` in loop is a very heavy operation